packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,356 @@
+
+/**
+ * Service for listening on backend connection changes.
+ */
+export const ConnectionStatusService = Symbol('ConnectionStatusService');
+export interface ConnectionStatusService {
+
+    /**
+     * The actual connection state.
+     */
+    readonly currentState: ConnectionStatus;
+
+    /**
+     * Clients can listen on connection status change events.
+     */
+    readonly onStatusChange: Event<ConnectionStatus>;
+
+}
+
+/**
+ * Connection status change event.
+ */
+export interface ConnectionStatus {
+
+    /**
+     * The current state of the connection.
+     */
+    readonly state: ConnectionState,
+
+    /**
+     * Optional health, percentage number.
+     */
+    readonly health?: number
+
+}
+
+/**
+ * The connection-status states.
+ */
+export enum ConnectionState {
+    ONLINE,
+    OFFLINE
+}
+
+@injectable()
+export class ConnectionStatusOptions {
+
+    static DEFAULT: ConnectionStatusOptions = {
+        requestTimeout: 1000,
+        retry: 5,
+        retryInterval: 1000,
+    };
+
+    /**
+     * Timeout for the HTTP GET request in milliseconds.
+     */
+    readonly requestTimeout: number;
+
+    /**
+     * Number of accepted timeouts.
+     */
+    readonly retry: number;
+
+    /**
+     * Retry interval in milliseconds.
+     */
+    readonly retryInterval: number;
+
+}
+
+@injectable()
+export class FrontendConnectionStatusService implements ConnectionStatusService, FrontendApplicationContribution {
+
+    static readonly MAX_RETRY_INTERVAL = 30000;
+
+    protected readonly statusChangeEmitter: Emitter<ConnectionStatus>;
+    private readonly endpointUrl: string;
+
+    private connectionState: ConnectionStateMachine;
+    private timer: number | undefined;
+    private retryInterval: number;
+
+    constructor(
+        @inject(ConnectionStatusOptions) @optional() protected readonly options: ConnectionStatusOptions = ConnectionStatusOptions.DEFAULT,
+        @inject(FrontendIdleService) protected idleService: FrontendIdleService,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) {
+        this.statusChangeEmitter = new Emitter<ConnectionStatus>();
+        this.endpointUrl = new Endpoint().getRestUrl().toString();
+        this.connectionState = new ConnectionStateMachine({ threshold: this.options.retry });
+        this.retryInterval = this.options.retryInterval;
+        this.idleService.onStatusChange(idle => {
+            if (idle) {
+                this.onStop();
+            } else {
+                this.onStart();
+            }
+        });
+    }
+
+    onStart() {
+        this.schedule(this.checkAlive.bind(this));
+        this.fireStatusChange(this.connectionState);
+        this.logger.info('Started backend connection status check.');
+    }
+
+    onStop() {
+        if (this.timer !== undefined) {
+            window.clearInterval(this.timer);
+            this.timer = undefined;
+            this.logger.info('Stopped backend connection status check.');
```

Comment:
`debug`? i don't think it is useful for an end user

Reply:
Leftover.